Java/jni migration

avifenesh opened the pull request:
<!--
Thanks for contributing to Valkey GLIDE!

Please make sure you are aware of our contributing guidelines [available
here](https://github.com/valkey-io/valkey-glide/blob/main/CONTRIBUTING.md)

-->

### Issue link

This Pull Request is linked to issue (URL): [REPLACE ME]

### Checklist

Before submitting the PR make sure the following are checked:

-   [ ] This Pull Request is related to one issue.
-   [ ] Commit message has a detailed description of what changed and why.
-   [ ] Tests are added or updated.
-   [ ] CHANGELOG.md and documentation files are updated.
-   [ ] Destination branch is correct - main or release
-   [ ] Create merge commit if merging release branch into main, squash otherwise.

jduo commented on java/client/src/main/java/glide/api/BaseClient.java:
```diff
@@ -333,28 +334,75 @@ public abstract class BaseClient
+                        if (pointer == null || pointer == 0) {
+                            return null;
+                        }
+                        // The pointer is now an ID to retrieve the object from the registry
```
```suggestion
                        // The pointer is an ID to retrieve the object from the registry
```

jduo commented on java/client/src/main/java/glide/api/BaseClient.java:
```diff
@@ -333,28 +334,75 @@ public abstract class BaseClient
+                        if (pointer == null || pointer == 0) {
+                            return null;
+                        }
+                        // The pointer is now an ID to retrieve the object from the registry
```
```suggestion
                        // The pointer is an ID to retrieve the object from the registry
```

jduo commented on java/client/src/main/java/glide/api/BaseClient.java:
```diff
@@ -333,28 +334,75 @@ public abstract class BaseClient
     }
 
-    /** Auxiliary builder which wraps all fields to be initialized in the constructor. */
+    /** Auxiliary builder which wraps all fields - matches UDS signature exactly */
```
Why highlight the matching of UDS signature? Is it for backwards compatibility reasons?

avifenesh replied:
work process, should be removed.

jduo commented on java/client/src/main/java/glide/api/BaseClient.java:
```diff
@@ -522,6 +615,23 @@ protected <T> T handleValkeyResponse(
+        // Handle DirectByteBuffer conversion for large responses (>16KB)
+        if (value instanceof java.nio.ByteBuffer) {
+            java.nio.ByteBuffer buffer = (java.nio.ByteBuffer) value;
+            byte[] bytes = new byte[buffer.remaining()];
```
Is it possible to write a fast-path that avoids this? deserializeMapFromBytes() just wraps the byte array into a ByteBuffer again.

jduo commented on java/client/src/main/java/glide/api/BaseClient.java:
```diff
@@ -522,6 +615,23 @@ protected <T> T handleValkeyResponse(
+            byte[] bytes = new byte[buffer.remaining()];
+            buffer.get(bytes);
+            if (classType == String.class && encodingUtf8) {
+                value = new String(bytes, java.nio.charset.StandardCharsets.UTF_8);
```
It looks like we can use StandardCharsets.UTF_8.decode(<ByteBuffer>) instead of copying to a byte array.

jduo commented on java/client/src/main/java/glide/api/BaseClient.java:
```diff
@@ -5053,6 +5215,54 @@ private Object convertByteArrayToGlideString(Object o) {
+                buffer.get(keyBytes);
+                Object key =
+                        encodingUtf8
+                                ? new String(keyBytes, java.nio.charset.StandardCharsets.UTF_8)
```
Use UTF_8.decode() directly on buffer.

jduo commented on java/client/src/main/java/glide/api/BaseClient.java:
```diff
@@ -5053,6 +5215,54 @@ private Object convertByteArrayToGlideString(Object o) {
+                    buffer.get(valueBytes);
+                    Object value =
+                            encodingUtf8
+                                    ? new String(valueBytes, java.nio.charset.StandardCharsets.UTF_8)
```
decode() instead of new String()

jduo commented on java/client/src/main/java/glide/api/GlideClusterClient.java:
```diff
@@ -156,9 +158,37 @@ public class GlideClusterClient extends BaseClient
+        return BaseClient.createClient(config, GlideClusterClient::new);
     }
 
+    /**
```
This comment looks like it is above the wrong method.

jduo commented on java/client/src/main/java/glide/api/GlideClusterClient.java:
```diff
@@ -1292,36 +1322,75 @@ public CompletableFuture<GlideString> randomKeyBinary() {
                 RandomKey, new GlideString[0], this::handleGlideStringOrNullResponse);
```
We should have a constant to represent an empty GlideString array to reduce allocations.

jduo commented on java/client/src/main/java/glide/api/GlideClusterClient.java:
```diff
@@ -1335,14 +1404,21 @@ private static final class NativeClusterScanCursor
             this.cursorHandle = cursorHandle;
-            this.isFinished = ClusterScanCursorResolver.FINISHED_CURSOR_HANDLE.equals(cursorHandle);
+            this.isFinished =
+                    ClusterScanCursorResolver.FINISHED_CURSOR_HANDLE != null
```
Why can this constant ever be null if it's a constant?

jduo commented on java/client/src/main/java/glide/ffi/resolvers/ScriptResolver.java:
```diff
@@ -1,25 +1,13 @@
-     * @param code The Lua script
-     * @return String representing the saved hash
-     */
     public static native String storeScript(byte[] code);
```
Why remove method comments?

jduo commented on java/client/src/main/java/glide/internal/AsyncRegistry.java:
```diff
@@ -0,0 +1,334 @@
+                    if (maxInflightRequests > 0) {
+                        java.util.concurrent.atomic.AtomicInteger clientCount =
+                                clientInflightCounts.get(clientHandle);
+                        if (clientCount != null) {
```
Can this ever be null?

avifenesh replied:
race condition defense

jduo commented on java/client/src/main/java/glide/internal/AsyncRegistry.java:
```diff
@@ -0,0 +1,334 @@
+                        : errorMessage;
+        RuntimeException ex;
+        switch (errorTypeCode) {
+            case 3: // Disconnect
```
Enum please

jduo commented on java/client/src/main/java/glide/internal/AsyncRegistry.java:
```diff
@@ -0,0 +1,334 @@
+    }
+
+    private static RuntimeException mapToTypedException(String errorMessage) {
+        String msg =
```
Can this be based on an error code at all? Using error message text is unreliable since localization can come into play.

jduo commented on java/client/src/main/java/glide/internal/GlideCoreClient.java:
```diff
@@ -0,0 +1,1100 @@
+                    Long, java.lang.ref.WeakReference<glide.api.BaseClient>>
+            clients = new java.util.concurrent.ConcurrentHashMap<>();
+
+    private static final byte[][] EMPTY_SCRIPT_PARAMS = new byte[0][];
```
Rename to be not script-specific and use elsewhere when we need an empty 2D byte array (eg subPattern).

jduo commented on java/client/src/main/java/glide/internal/GlideCoreClient.java:
```diff
@@ -0,0 +1,1100 @@
+            result.clusterMode(false);
+
+            if (standaloneConfig.getDatabaseId() != null) {
+                result.databaseId(standaloneConfig.getDatabaseId());
```
DatabaseId is now part of BaseClientConfiguration and should apply to both cluster and standalone modes. Please rebase or merge if that's not the case.

jduo commented:
Do we still need protobuf?

jduo commented on java/client/src/main/java/glide/internal/protocol/BinaryCommandRequest.java:
```diff
@@ -0,0 +1,143 @@
+    }
+
+    /** Serialize to binary format for JNI transmission while preserving binary data */
+    public byte[] toBytes() {
```
There looks to be alot of duplication between this and BatchRequest#writeCommand().

jduo commented on java/client/src/main/java/glide/internal/protocol/BinaryCommandRequest.java:
```diff
@@ -0,0 +1,143 @@
+        }
+    }
+
+    private void writeVarInt(ByteArrayOutputStream baos, int value) throws IOException {
```
This method is duplicated in BatchRequest.

jduo commented on java/client/src/main/java/glide/internal/protocol/CommandRequest.java:
```diff
@@ -0,0 +1,125 @@
+            ByteArrayOutputStream baos = new ByteArrayOutputStream();
+
+            // Write command name length and data
+            byte[] cmdBytes = commandName.getBytes(StandardCharsets.UTF_8);
```
Is it worth having a lookup table that caches command names as bytes? It seems likely that users will reuse the same commands in the same program.

jduo commented on java/client/src/main/java/glide/managers/CommandManager.java:
```diff
@@ -31,343 +29,860 @@
+        byte[] bytes = new byte[dup.remaining()];
+        dup.get(bytes);
+        if (expectUtf8Response) {
+            return new String(bytes, java.nio.charset.StandardCharsets.UTF_8);
```
Use StandardCharsets.UTF_8.decode(ByteBuffer) instead of copying to byte array temporary.

avifenesh commented on java/client/src/main/java/glide/api/BaseClient.java:
```diff
@@ -5543,4 +5753,52 @@ public CompletableFuture<Long> wait(long numreplicas, long timeout) {
+     * Internal method for enqueueing PubSub messages from native callback. This is called by the
+     * native layer when PubSub messages are received.
+     */
+    public void __enqueuePubSubMessage(PubSubMessage message) {
```
for cross-package access from glide.internal.GlideCoreClient to glide.api.BaseClient

avifenesh replied:
we use it for glidecoreclient as well, so we need cross package visibility.

avifenesh commented on java/client/src/main/java/glide/api/BaseClient.java:
```diff
@@ -790,6 +900,18 @@ protected Map<String, Object> handleLcsIdxResponse(Map<String, Object> response)
+        Map<String, Object> converted = new LinkedHashMap<>(Math.max(response.size(), 1));
+        for (Map.Entry<GlideString, Object> entry : response.entrySet()) {
+            GlideString key = entry.getKey();
+            converted.put(key != null ? key.toString() : null, entry.getValue());
```
base on my understanding LinkedHashMap allow does allow null, inheriting this behavior from hashmap. But correct me if im wrong. It generally less possible that key will be null because of the contract with Valkey, but im not sure whether it is common in Java to "trust" the return type. Its sounds like its a wrong behavior to not check at all.

avifenesh replied:
But i guess we should fail fast here since if null returned something wrong happen.

avifenesh replied:
changed to fail fast.

avifenesh commented on java/client/src/main/java/glide/api/BaseClient.java:
```diff
@@ -2244,6 +2376,29 @@ public CompletableFuture<GlideString> scriptShow(@NonNull GlideString sha1) {
+    }
+
+    public CompletableFuture<String> scriptFlush() {
+        return commandManager.submitNewCommand(ScriptFlush, new String[0], this::handleStringResponse);
```
good one, thanks, done over the file in general.

avifenesh commented:
> Do we still need protobuf?

@jduo I apologize. I removed Protobuf for the manually made protocol, then added it back and removed the manual but missed a few changes, and mistakenly, for some reason, I kept maintaining the protocol files and wiring the execute calls, but they are unused. I would with pleasure remove the Protobuf completely, but it was pushed back because of the size of the changes.

jduo commented on java/client/src/main/java/glide/api/models/commands/scan/ClusterScanCursor.java:
```diff
@@ -86,7 +86,17 @@ public void releaseCursorHandle() {
 
     /** Creates an empty cursor to be used in the initial {@link GenericClusterCommands#scan} call. */
-    static ClusterScanCursor initalCursor() {
+    static ClusterScanCursor initialCursor() {
```
This is unfortunately a breaking change so we technically should keep initalCursor() and mark it deprecated.

jduo commented on java/client/src/main/java/glide/internal/AsyncRegistry.java:
```diff
@@ -0,0 +1,261 @@
+                            int remaining = clientCount.decrementAndGet();
+                            // Clean up the entry when no more inflight requests
+                            // This prevents memory leak for inactive clients
+                            if (remaining == 0) {
```
This doesn't happen atomically. There's a possible race condition if you get new activity on the client after decrementAndGet() runs, but before you reach this.

We can use clientInflightCounts.compute() to do this atomically. Have the lambda return null if decrementAndCount() returns zero and it'll remove the key.

avifenesh replied:
done!

avifenesh commented on java/client/src/main/java/glide/internal/GlideCoreClient.java:
```diff
@@ -0,0 +1,970 @@
+
+    // Removed blocking command detection - Rust handles all timeout logic
+
+    private static String formatSpanName(String commandName) {
```
good catch

avifenesh commented:
@codex

jeremyprime commented:
GH actions are failing to compile due to `AuthenticationInfo` struct now requiring `iam_config`.  I was able to build and run tests locally by updating `jni_client.rs` (assuming adding a `None` value doesn't cause any other issues):
```
            (Some(user), Some(pass)) => Some(AuthenticationInfo {
                username: Some(user),
                password: Some(pass),
                iam_config: None,
            }),
            (None, Some(pass)) => Some(AuthenticationInfo {
                username: None,
                password: Some(pass),
                iam_config: None,
            }),
```

jeremyprime commented on java/client/build.gradle:
```diff
@@ -164,13 +131,8 @@ tasks.register('buildWithProto') {
     finalizedBy 'build'
 }
 
-tasks.register('testFfi', Test) {
```
Should also update `DEVELOPER.md` which references running FFI tests.

avifenesh replied:
done

avifenesh commented on java/client/build.gradle:
```diff
@@ -7,28 +7,22 @@ plugins {
-    implementation group: 'io.netty', name: 'netty-transport-native-epoll', version: '4.1.121.Final', classifier: 'linux-aarch_64'
-    implementation group: 'io.netty', name: 'netty-transport-native-kqueue', version: '4.1.121.Final', classifier: 'osx-aarch_64'
-    implementation group: 'io.netty', name: 'netty-transport-native-kqueue', version: '4.1.121.Final', classifier: 'osx-x86_64'
+    compileOnly 'com.github.spotbugs:spotbugs-annotations:4.8.6'
```
done

avifenesh commented:
> GH actions are failing to compile due to `AuthenticationInfo` struct now requiring `iam_config`. I was able to build and run tests locally by updating `jni_client.rs` (assuming adding a `None` value doesn't cause any other issues):
> 
> ```
>             (Some(user), Some(pass)) => Some(AuthenticationInfo {
>                 username: Some(user),
>                 password: Some(pass),
>                 iam_config: None,
>             }),
>             (None, Some(pass)) => Some(AuthenticationInfo {
>                 username: None,
>                 password: Some(pass),
>                 iam_config: None,
>             }),
> ```

Thanks for the review—I applied the general implementation to unblock the CI in general.
Those are unrelated to the PR, they are new impl but to get green CI and since the PR is reimplementing many code pieces, implemented all the functionality.
